File: app/src/main/java/com/amplifyframework/datastore/sample/RemotePresenter.java
package com.amplifyframework.datastore.sample;

import com.amplifyframework.datastore.appsync.ModelWithMetadata;
import com.amplifyframework.datastore.generated.model.Post;

import java.util.Locale;

import io.reactivex.disposables.CompositeDisposable;

final class RemotePresenter implements RemotePresentation.Presenter {
    private final RemotePresentation.ApiInteractor interactor;
    private final RemotePresentation.View view;
    private final CompositeDisposable ongoingOperations;

    RemotePresenter(RemotePresentation.ApiInteractor interactor, RemotePresentation.View view) {
        this.interactor = interactor;
        this.view = view;
        this.ongoingOperations = new CompositeDisposable();
    }

    @Override
    public void createRemotePost() {
        ongoingOperations.clear();
        ongoingOperations.add(
            interactor.createRandom()
                .subscribe(
                createdItem ->
                    view.displayRemoteLogLine(LogLine.create(title(createdItem), details(createdItem))),
                failure ->  view.displayRemoteLogLine(LogLine.error("Failed to create a post.", failure.getMessage()))
            )
        );
    }

    @Override
    public void updateRemotePosts() {
        ongoingOperations.clear();
        ongoingOperations.add(
            interactor.list()
                .filter(mwm -> !Boolean.TRUE.equals(mwm.getSyncMetadata().isDeleted()))
                .flatMapSingle(interactor::update)
                .subscribe(
                    updatedItem -> view.displayRemoteLogLine(LogLine.update(title(updatedItem), details(updatedItem))),
                    failure -> view.displayRemoteLogLine(LogLine.error("Failed to update a post.", failure.getMessage()))
                )
        );
    }

    @Override
    public void deleteRemotePosts() {
        ongoingOperations.clear();
        ongoingOperations.add(
            interactor.list()
                .filter(mwm -> !Boolean.TRUE.equals(mwm.getSyncMetadata().isDeleted()))
                .flatMapSingle(interactor::delete)
                .subscribe(
                    deletedItem -> view.displayRemoteLogLine(LogLine.delete(title(deletedItem), details(deletedItem))),
                    failure -> view.displayRemoteLogLine(LogLine.error("Failed to delete a post.", failure.getMessage()))
                )
        );
    }

    @Override
    public void listRemotePosts() {
        ongoingOperations.clear();
        ongoingOperations.add(
            interactor.list()
                .filter(mwm -> !Boolean.TRUE.equals(mwm.getSyncMetadata().isDeleted()))
                .subscribe(
                    foundItem -> view.displayRemoteLogLine(LogLine.query(title(foundItem), details(foundItem))),
                    failure -> view.displayRemoteLogLine(LogLine.error("Failure to list items.", failure.getMessage()))
                )
        );
    }

    @Override
    public void clearRemoteLogs() {
        view.clearRemoteLineItems();
    }

    private static String title(ModelWithMetadata<Post> modelWithMetadta) {
        return modelWithMetadta.getModel().getTitle();
    }

    private static String details(ModelWithMetadata<Post> modelWithMetadata) {
        String hash = modelWithMetadata.getSyncMetadata().getId().substring(0, 7);
        String version = String.valueOf(modelWithMetadata.getSyncMetadata().getVersion());
        Boolean deleted = modelWithMetadata.getSyncMetadata().isDeleted();
        return String.format(Locale.US, "%s, %s, %b", hash, version, deleted);
    }
}

File: app/src/main/java/com/amplifyframework/datastore/sample/RemotePresenter.kt
package com.amplifyframework.datastore.sample

import com.amplifyframework.datastore.appsync.ModelWithMetadata
import com.amplifyframework.datastore.generated.model.Post
import com.amplifyframework.datastore.sample.RemotePresentation.ApiInteractor
import com.amplifyframework.datastore.sample.RemotePresentation.Presenter
import com.amplifyframework.datastore.sample.RemotePresentation.View

import io.reactivex.disposables.CompositeDisposable

import java.util.Locale

internal class RemotePresenter(private val interactor: ApiInteractor, private val view: View) : Presenter {
    private val ongoingOperations: CompositeDisposable = CompositeDisposable()

    override fun createRemotePost() {
        ongoingOperations.clear()
        ongoingOperations.add(interactor.createRandom().subscribe(
            { view.displayRemoteLogLine(LogLine.create(title(it), details(it))) },
            { view.displayRemoteLogLine(LogLine.error("Failed to create a post.", it.message!!)) }
        ))
    }

    override fun updateRemotePosts() {
        ongoingOperations.clear()
        ongoingOperations.add(interactor.list()
            .filter { true != it.syncMetadata.isDeleted }
            .flatMapSingle { interactor.update(it) }
            .subscribe(
                { view.displayRemoteLogLine(LogLine.update(title(it), details(it))) },
                { view.displayRemoteLogLine(LogLine.error("Failed to update a post.", it.message!!)) }
            )
        )
    }

    override fun deleteRemotePosts() {
        ongoingOperations.clear()
        ongoingOperations.add(interactor.list()
            .filter { true != it.syncMetadata.isDeleted }
            .flatMapSingle { interactor.delete(it) }
            .subscribe(
                { view.displayRemoteLogLine(LogLine.delete(title(it), details(it))) },
                { view.displayRemoteLogLine(LogLine.error("Failed to delete a post.", it.message!!)) }
            )
        )
    }

    override fun listRemotePosts() {
        ongoingOperations.clear()
        ongoingOperations.add(interactor.list()
            .filter { true != it.syncMetadata.isDeleted }
            .subscribe(
                { view.displayRemoteLogLine(LogLine.query(title(it), details(it))) },
                { view.displayRemoteLogLine(LogLine.error("Failure to list items.", it.message!!)) }
            )
        )
    }

    override fun clearRemoteLogs() {
        view.clearRemoteLineItems()
    }

    companion object {
        private fun title(modelWithMetadata: ModelWithMetadata<Post>): String {
            return modelWithMetadata.model.title
        }

        private fun details(modelWithMetadata: ModelWithMetadata<Post>): String {
            val hash = modelWithMetadata.syncMetadata.id.substring(0, 7)
            val version = modelWithMetadata.syncMetadata.version.toString()
            val deleted = modelWithMetadata.syncMetadata.isDeleted
            return String.format(Locale.US, "%s, %s, %b", hash, version, deleted)
        }
    }
}
